sources/api/netdog/src/networkd_status.rs

```diff
@@ -0,0 +1,305 @@
+        match self.addresses.len().cmp(&1) {
+            Ordering::Less => error::NoIpAddressSnafu {
+                interface: self.name.clone(),
+            }
+            .fail(),
+            Ordering::Equal => Ok(self.addresses[0]),
+            Ordering::Greater => {
+                // Find IPv4 first
+                for addr in self.addresses.iter() {
+                    if addr.is_ipv4() {
+                        return Ok(*addr);
+                    }
+                }
+                // No IPv4 addresses, then attempt to return the first IPv6, otherwise return error
+                match self.addresses.first() {
+                    Some(addr) => Ok(*addr),
+                    None => error::NoIpAddressSnafu {
+                        interface: self.name.clone(),
+                    }
+                    .fail(),
+                }
+            }
+        }
+    }
+}
+
+mod error {
+    use crate::interface_id::InterfaceId;
+    use snafu::Snafu;
+    use std::{io, string::FromUtf8Error};
+
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility(pub(crate)))]
+    pub(crate) enum Error {
+        #[snafu(display("Failed to run 'networkctl': {}", source))]
+        NetworkctlExecution { source: io::Error },
+
+        #[snafu(display("'networkctl' failed: {}", stderr))]
+        FailedNetworkctl { stderr: String },
+
+        // Consider maybe shifting to length only for this error "Expected a Vec of length 4 but it was {}", v.len()
+        #[snafu(display("Failed to parse IP Address: {:?} {}", input, msg))]
+        BadIpAddress { input: Vec<u8>, msg: String },
+
+        #[snafu(display("Failed to parse IP Address from networkctl status"))]
+        BadIpAddressVec,
+
+        #[snafu(display("No IP Address for Primary Interface: {:?}", interface))]
+        NoIpAddress { interface: InterfaceId },
+
+        #[snafu(display("Failed to parse 'networkctl' output: {}", source))]
+        NetworkctlParsing { source: FromUtf8Error },
+
+        #[snafu(display("Failed to deserialize 'networkctl' output: {}", source))]
+        NetworkctlDeserialize { source: serde_json::Error },
+
+        #[snafu(display("No IP Address found for Primay Interface: {}", source))]
+        NoPrimaryIpAddress { source: io::Error },
+    }
+}
+
+pub(crate) use error::Error as NetworkDStatusError;
+type Result<T> = std::result::Result<T, NetworkDStatusError>;
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::fs;
+    use std::net::Ipv4Addr;
+    use std::path::Path;
+    use std::path::PathBuf;
+
+    fn test_data() -> PathBuf {
+        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("test_data")
+    }
+
+    fn networkd_config() -> PathBuf {
+        test_data().join("networkd")
+    }
+
+    fn read_output_file<P>(path: P) -> String
+    where
+        P: AsRef<Path>,
+    {
+        let path = path.as_ref();
+        let path_str = fs::read_to_string(path).unwrap();
+
+        path_str
+    }
```

Comment:
My editor flags this as "returning the result of a `let` binding from a block". There are a few other lints that apply to the test code.

```suggestion
    fn read_output_file<P>(path: P) -> String
    where
        P: AsRef<Path>,
    {
        let path = path.as_ref();
        fs::read_to_string(path).unwrap()
    }
```

Reply:
Oh, this is fair, I didn't look too closely at the test code and my editor isn't flagging this so its worth me following up to see why. I'll improve this.